payments-core/src/main/java/com/stripe/android/googlepaylauncher/GooglePayLauncherActivity.kt

```diff
@@ -0,0 +1,216 @@
+import android.os.Bundle
+import androidx.activity.viewModels
+import androidx.appcompat.app.AppCompatActivity
+import androidx.lifecycle.lifecycleScope
+import com.google.android.gms.wallet.AutoResolveHelper
+import com.google.android.gms.wallet.PaymentData
+import com.google.android.gms.wallet.PaymentDataRequest
+import com.google.android.gms.wallet.PaymentsClient
+import com.stripe.android.model.PaymentMethodCreateParams
+import com.stripe.android.model.StripeIntent
+import com.stripe.android.view.AuthActivityStarterHost
+import kotlinx.coroutines.launch
+import org.json.JSONObject
+
+/**
+ * [GooglePayLauncherActivity] is used to return the result of a Google Pay operation.
+ * The activity's result will be a [GooglePayLauncher.Result].
+ *
+ * [GooglePayLauncherActivity] will:
+ * 1. Fetch the [StripeIntent]
+ * 2. Verify that Google Pay is available
+ * 3. Create a [PaymentDataRequest](https://developers.google.com/pay/api/web/reference/request-objects#PaymentDataRequest)
+ * 4. Confirm the [StripeIntent] using the Google Pay token
+ *
+ * Use [GooglePayLauncherContract] to start [GooglePayLauncherActivity].
+ *
+ * See [Troubleshooting](https://developers.google.com/pay/api/android/support/troubleshooting)
+ * for a guide to troubleshooting Google Pay issues.
+ */
+internal class GooglePayLauncherActivity : AppCompatActivity() {
+    private val paymentsClient: PaymentsClient by lazy {
+        PaymentsClientFactory(this).create(args.config.environment)
+    }
+    private val viewModel: GooglePayLauncherViewModel by viewModels {
+        GooglePayLauncherViewModel.Factory(
+            application,
+            args
+        )
+    }
+
+    private lateinit var args: GooglePayLauncherContract.Args
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        if (Build.VERSION.SDK_INT != Build.VERSION_CODES.O) {
+            // In Oreo, Activities where `android:windowIsTranslucent=true` can't request
+            // orientation. See https://stackoverflow.com/a/50832408/11103900
+            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
+        }
+
+        overridePendingTransition(0, 0)
+        setResult(
+            RESULT_OK,
+            Intent().putExtras(
+                GooglePayLauncherResult.Canceled.toBundle()
+            )
+        )
+
+        val nullableArgs = GooglePayLauncherContract.Args.fromIntent(intent)
+        if (nullableArgs == null) {
+            finishWithResult(
+                GooglePayLauncher.Result.Failed(
+                    RuntimeException(
+                        "GooglePayLauncherActivity was started without arguments."
+                    )
+                )
+            )
+            return
+        }
+        args = nullableArgs
+
+        viewModel.googlePayResult.observe(this) { googlePayResult ->
+            googlePayResult?.let(::finishWithResult)
+        }
+
+        if (!viewModel.hasLaunched) {
```

Comment:
since we only want to call `createPaymentDataRequest` once within a lifecycle of a `viewModel`, I think this check can be encapsulate inside `createPaymentDataRequest`, within which `viewModelScope` is used to launch #getStripeIntent coroutine

Reply:
I could do that but I'm not clear what the return type would be.